src/math/maxBy.ts

```diff
@@ -0,0 +1,36 @@
+/**
+ * Returns the element of the specified array that has the maximum value, according to the specified selector.
+ *
+ * This function takes an array of elements and a selector function, and returns the element with the maximum value according to the selector.
+ *
+ * @param {T[]} elements - An array of elements to be compared.
+ * @param {(element: T) => number} selector - A function that takes an element and returns a number.
+ * @returns {T} The element with the maximum value according to the selector.
+ *
+ * @example
+ * const people = [
+ *  { name: 'Mark', age: 25 },
+ *  { name: 'Nunu', age: 30 },
+ *  { name: 'Overmars', age: 20 }
+ *  ];
+ *  const result = maxBy(people, person => person.age);
+ *  // result will be { name: 'Nunu', age: 30 }
+ */
+export function maxBy<T>(elements: T[], selector: (element: T) => number): T | undefined {
```

Comment:
```suggestion
export function maxBy<T>(elements: T[], selector: (element: T) => number): T {
```
I believe returning `undefined` is stricter and less error-prone.

However, following TypeScript's convention that `arr[0]` is not nullable even if `arr` is empty (unless [noUncheckedIndexedAccess](https://www.typescriptlang.org/tsconfig/#noUncheckedIndexedAccess) is on) is also important for consistency -- that's why we chose `sample` to not return nullable even if it can return `undefined`.

For the sake of consistency, I believe we should not return `undefined` here.

What do you think?

Reply:
> However, following TypeScript's convention that arr[0] is not nullable even if arr is empty

I tested this on TS Playground. And the result is below.

```typescript
const array: number[] = [];
console.log(array[0]);

// [LOG]: undefined 
```

I think this log says it's null(or undefined) when point first element of empty array.

<img width="1624" alt="스크린샷 2024-06-17 오후 2 32 28" src="https://github.com/toss/es-toolkit/assets/54518925/f6618a30-29d6-48fe-b1ac-76a059328687">

--------------------------

I think usage without undefined makes compilcated usecase. Let's see below example code.

## When we use undefined

```ts
  const maxElement = maxBy<number>([]) ?? 0;
```

## When we use Exception

```ts
let maxElement;
try {
  maxElement = maxBy<number>([]);
} catch (error) {
  // TODO: Logic when error occurred
}

const result = maxElement;
```

I think above one is more simple. What do you think?

--------------------------

But for the sake of consistency in this library, I also think it should not return undefined. Thanks for your advice.

Reply:
<img width="311" alt="image" src="https://github.com/toss/es-toolkit/assets/3102175/db6c87cf-39fc-409c-9e0b-3b39313250a0">

Oh, what I thought was that here `item` is not nullable (in static types) even if `array` is empty.

Reply:
So I thought dealing with empty arrays is what users of `es-toolkit` have to handle... 
(We do not guarantee that the array is not empty using types)

Reply:
I see... It seems that usual convention in typescript. It's so hard to admit, but now I'm understand what you wrote. I'll change to non-nil type! Thanks for check it.

Reply:
After modification, all test cases pass well 👍🏻